src/quickcheck.rs

```diff
@@ -0,0 +1,264 @@
+//! Implementations of the `quickcheck::Arbitrary` trait, which enables users to
+//! write tests like this, and have test values provided automatically:
+//!
+//! ```
+//! use quickcheck::quickcheck;
+//! use time::Date;
+//!
+//! struct DateRange {
+//!     from: Date,
+//!     to: Date,
+//! }
+//!
+//! impl DateRange {
+//!     fn new(from: Date, to: Date) -> Result<Self, ()> {
+//!         Ok(DateRange { from, to })
+//!     }
+//! }
+//!
+//! quickcheck! {
+//!     fn date_range_is_well_defined(from: Date, to: Date) -> bool {
+//!         let r = DateRange::new(from, to);
+//!         if from <= to {
+//!             r.is_ok()
+//!         } else {
+//!             r.is_err()
+//!         }
+//!     }
+//! }
+//! ```
+//!
+//! An implementation for [`Instant`] is intentionally omitted since its values
+//! are only meaningful in relation to a [`Duration`], and obtaining an
+//! [`Instant`] from a [`Duration`] is very simple anyway.
+
+use crate::{
+    date::{Date, MAX_YEAR, MIN_YEAR},
+    duration::Duration,
+    offset_date_time::OffsetDateTime,
+    primitive_date_time::PrimitiveDateTime,
+    time_mod::Time,
+    utc_offset::UtcOffset,
+    util::days_in_year,
+    weekday::Weekday,
+};
+use quickcheck::{Arbitrary, Gen};
+use rand::Rng;
+use std::{
+    cmp::{max, min},
+    convert::TryFrom,
+};
+
+impl Arbitrary for Date {
+    fn arbitrary<G>(g: &mut G) -> Self
+    where
+        G: Gen,
+    {
+        let y_size: i32 = i32::try_from(g.size()).unwrap_or(i32::max_value());
+        let y: i32 = if y_size == 0 {
+            0
+        } else {
+            g.gen_range(max(MIN_YEAR, -y_size), min(MAX_YEAR, y_size))
+        };
+
+        let o_size: u16 = min(
+            u16::try_from(g.size()).unwrap_or(u16::max_value()),
+            days_in_year(y),
+        );
+        let o = g.gen_range(1, max(2, o_size + 1));
+
+        Self::from_yo_unchecked(y, o)
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        let self_y = self.year();
```

Comment:
This could probably just be called `year`.

Reply:
Yep, agreed.